lib/cisco_node_utils/span_session.rb

```diff
@@ -0,0 +1,123 @@
+# Insert appropriate license here
+require_relative 'node_util'
+require_relative 'interface'
+
+module Cisco
+  # node utils class for SPAN (switchport analyzer) sessions
+  class SpanSession < NodeUtil
+    attr_reader :session_id
+
+    def initialize(session_id, instantiate=true)
+      validate_args(session_id.to_i)
+      create if instantiate
+    end
+
+    def self.sessions
+      hash = {}
+      all = config_get('span_session', 'all_sessions')
+      return hash if all.nil?
+
+      all.each do |id|
+        hash[id] = SpanSession.new(id, false)
+      end
+      hash
+    end
+
+    def validate_args(session_id)
+      fail TypeError unless session_id.is_a?(Integer)
+      @session_id = session_id
+    end
+
+    def create
+      config_set('span_session', 'create', id: @session_id)
+    end
+
+    def destroy
+      config_set('span_session', 'destroy', id: @session_id)
+    end
+
+    def description
+      config_get('span_session', 'description', id: @session_id)
+    end
+
+    def description=(val)
+      val = val.to_s
+      if val.empty?
+        config_set('span_session', 'description', id: @session_id,
+                    state: 'no', description: '')
+      else
+        config_set('span_session', 'description', id: @session_id,
+                    state: '', description: val)
+      end
+    end
+
+    def destination
+      config_get('span_session', 'destination', id: @session_id)
+    end
+
+    def destination=(int)
+      # fail if int is not a valid interface
+      fail TypeError unless Interface.interfaces.key?(int)
+      config_set('span_session', 'destination', id: @session_id, intf_name: int)
+    end
+
+    def session_id
+      config_get('span_session', 'session_id')
+    end
+
+    def session_id=(id)
+      fail TypeError unless id.is_a?(Integer)
+      config_set('span_session', 'session_id', id: id, state: '')
+    end
+
+    def shutdown
+      config_get('span_session', 'shutdown', id: @session_id)
+    end
+
+    def shutdown=(bool)
+      fail TypeError unless bool.is_a?(Boolean)
+      config_set('span_session', 'shutdown', id: @session_id, shutdown: bool)
+    end
+
+    def source_interface
+      config_get('span_session', 'source_interface', id: @session_id)
+    end
+
+    def source_interface=(sources)
```

Comment:
@tomcooperca I ran the tests with your latest diffs but still see a couple of failures.

```
TestSpanSession#test_session_source_vlans = 1.99 s = E
TestSpanSession#test_remove_session = 1.54 s = .
TestSpanSession#test_session_type = 1.56 s = F
TestSpanSession#test_create_session_invalid_id = 1.48 s = .
TestSpanSession#test_session_description = 1.61 s = .
TestSpanSession#test_session_destination_int = 2.76 s = .
TestSpanSession#test_create_new_session = 1.57 s = .
TestSpanSession#test_session_source_interface = 2.89 s = E

Finished in 20.339367s, 0.3933 runs/s, 0.4917 assertions/s.

  1) Error:
TestSpanSession#test_session_source_vlans:
TypeError: TypeError
    /nobackup/mwiebe/NU/tom/cisco-network-node-utils/lib/cisco_node_utils/span_session.rb:104:in `block in source_vlan='
    /nobackup/mwiebe/NU/tom/cisco-network-node-utils/lib/cisco_node_utils/span_session.rb:103:in `each_pair'
    /nobackup/mwiebe/NU/tom/cisco-network-node-utils/lib/cisco_node_utils/span_session.rb:103:in `source_vlan='
    tests/test_span_session.rb:83:in `test_session_source_vlans'


  2) Failure:
TestSpanSession#test_session_type [tests/test_span_session.rb:53]:
Expected: "local"
  Actual: "erspan-source"


  3) Error:
TestSpanSession#test_session_source_interface:
TypeError: TypeError
    /nobackup/mwiebe/NU/tom/cisco-network-node-utils/lib/cisco_node_utils/span_session.rb:91:in `block in source_interface='
    /nobackup/mwiebe/NU/tom/cisco-network-node-utils/lib/cisco_node_utils/span_session.rb:89:in `each_pair'
    /nobackup/mwiebe/NU/tom/cisco-network-node-utils/lib/cisco_node_utils/span_session.rb:89:in `source_interface='
    tests/test_span_session.rb:70:in `test_session_source_interface'

8 runs, 10 assertions, 1 failures, 2 errors, 0 skips
```

If you are working on additional changes, I will wait for new diffs.  If you want me to investigate and make modifications let me know.  Thanks!